tools/clang/test/HLSLFileCheck/hlsl/template/missing_typename.hlsl

```diff
@@ -7,12 +7,12 @@
 // PRESENT: define void @VSMain() {
 
 template <typename T>
+typedef T value_type;
 struct Vec2 {
     T x, y;
     T Length() {
         return sqrt(x*x + y*y);
     }
-    typedef T value_type;
```

Comment:
Did you mean to change this? I don't think this test case should be getting changed. If something in your change requires this change to be made, that is concerning.

Reply:
The test case did need changes; however, it looks like I've included the wrong version in the patch - this will be updated soon. In the meantime the correct text can be found in the [attached text file](https://github.com/microsoft/DirectXShaderCompiler/files/7629930/missing_typename.txt).

As for why it needed to be changed, the original test was not valid and was only passing because of the very issue being dealt with here. The example code in the original issue would fail because there was an early exit in the parsing of the struct with templates enabled - as a result, not all contents of the struct would be processed, and so trying to access them later in the code would cause an error. On the flip side, for the above test case, the early exit actually prevented the compiler from catching an error (nested typedef declarations are illegal) so the test had to be modified to be able to test the same concept without the nested typedef. Once again, thought, this is the wrong version of the test case, and the correct one will be added soon. @tcorringham can better explain the amended test case if need be.

Reply:
I'd really like to understand why you needed changes in this test file. That test file's code is valid C++, and the attached text file is not at all equivalent.

Reply:
This set of test cases is written in HLSL, which does not have all of the support and capabilities of C++ just yet. From my understanding, nested typedefs are excluded from the current version of the compiler by design, as it has been explicitly enumerated as an error. As such it is fairly certain that the original test case should not have been usable and requires changes.

I'll defer to Tim for any explanation on the equivalency of the tests - he authored both the original and amended versions of this test case, so he has a better understanding of what particular scenario it is meant to verify.

Reply:
Just some background to this test case. I added this test as part of a fix for #3556 - where a cascade of unhelpful errors were produced rather than the expected diagnostic as in C++.
The test was taken from the example C++ code in the issue report - and was intended only to demonstrate the different behaviour before and after the fix for #3556. The use of the nested typedef is not supported in HLSL, although previously no diagnostic was produced in this case.
The fix for #3731 does elicit an error for the nested typedef - I suggested removing the typedef to preserve this test as a valid test for the #3556 fix, as it is incidental to that. The amended test still demonstrates the expected before/after differences.
So I suggest that this amended test is valid - although it could be argued that the lack of a diagnostic for the nested typdef previously was a bug.

Reply:
Before this change, the test case is valid C++, that should produce an error of some kind (useful or otherwise) in HLSL. With this change or the proposed change in the text file, the test case is invalid code in C++ and HLSL. IMO, that reduces the usefulness of the test case.

Clearly the test is wrong, in that it had a success case where it allowed the code to compile, (no arguments fixing that), and maybe the test case produces a different error (useful or otherwise, I'm okay with that). But changing the valid C++ to invalid C++ to produce the same error and allow strange code to compile seems like the wrong change to the test case.

Reply:
So, I ran the test locally (unmodified), and can confirm that the different error message produced is actually the error message we want to produce (HLSL doesn't support nested typedefs). I think the only problem with this test case is that it checks for the wrong result.